docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -221,24 +221,44 @@ Horizontal Pod Autoscalers also need to be considered and removed.
 
 #### If components in the __Source Seed__ cannot reliably read who the leader is from the Garden cluster
+**Using timestamps in the ETCD backup entries:**
 
-Currently we have come up with two ideas to handle this case:
+The leader election depends on two header files:
+1. Header file in the `Shoot`'s backup parent directory indicating who is the master and when the migration happened as a timestamp.
+2. Header file in each snapshot folder with timestamps.
 
-**DNS leader election:** A DNS TXT entry with TTL=60s and value seed='__Source Seed__' is used. The record is created and maintained by the Gardener Controller Manager (by using the DNS Controller Manager and its DNSEntry resource). When a control plane migration is detected, the Gardener Controller Manager changes the value of the DNS Entry to seed='__Destination Seed__' and waits for 2*TTL + 1 = 121 seconds to ensure that the change is propagated to all controllers in the old seed. We rely on the fact that DNS is highly available (100% for AWS Route53) and that control plane components in the __Source Seed__ can see the changes.
+When migration is triggered:
+1. The ETCD Druid in the __Destination Seed__ is notified of the start of the migration. This "notification" can happen via the `Cluster` resource with an additional field indicating when the migration was actually triggered as a timestamp.
+```
+spec:
+  migrationTimestamp: "2018-11-18T15:14:57Z"
+```
+2. The ETCD Druid adds this migration timestamp to the header file inside the `Shoot`'s backup parent directory together with the __Destination Seed's__ name to indicate who the master is.
+3. When the `Shoot`'s ETCD pod is launched in the __Destination Seed__ the backup-restore sidecar checks the available backups. It restores the ETCD data from the snapshot with the latest timestamp (older then the migration timestamp).
+4. The `Shoot`'s ETCD `backup-restore` sidecar in the __Source Seed__ checks if the `Seed` name in the backup parent directory header matches the current __Source Seed__.
+    - If it does not match, it does not take a snapshot and should kill the ETCD container.
+    - If it matches - migration has not been triggerd a snapshot can be made.
+5. Once the ETCD container dies, all components which depend on it (e.g. the Kube APIServer,  Machine Controller Manager and so on) will also stop working.
+
+However, additional work has to be done for other components which do not rely on the `Shoot`'s ETCD. Following is one proposal to handle this case:
+
+The `Cluster` resource has an additional timestamp field which is updated either by the ETCD Druid or the `backup-restore` sidecar every `X` seconds. If the ETCD Druid is used to update the timestamp it must monitor the health of the `backup-restore` sidecar. However, if `backup-restore` is used - it must have knowledge of kubernetes.
```

Comment:
The `Cluster` resource is quite large, and updating it periodically will stress the seed's etcd. Probably a `Lease` object is more applicable here?

Reply:
Yep,I was thinking about having a separate resource to handle this a bit after I updated the GEP.

Reply:
Can't the `Etcd` resource be used here? It isn't that large and is quite relevant to the topic.

Reply:
> The `Cluster` resource has an additional timestamp field which is updated either by the ETCD Druid or the `backup-restore` sidecar every `X` seconds.

Are we proposing a kind of lease or health information via the `Cluster` resource? If so, why is this needed?

Reply:
>Are we proposing a kind of lease or health information via the Cluster resource? If so, why is this needed?

This lease information is necessary for the extension controllers which means that they would have to read the `Etcd` resource. Additionally, I think that having a different resource might be better if in the future we decide to use some other form of leader election.

Reply:
👍 In the https://github.com/gardener/gardener/pull/3066#discussion_r518661666 below, I have argued against leader _election_ because technically we do not need it. We only need leader _detection_. But there I assumed that we will continue to use `Cluster` and `Etcd` resources to detect leaders. But with that approach, we can use some other resource, of course. Only not the `Lease` resource, though. Do let me know what you think.